apache2/waf_logging/waf_log_util.cc

```diff
@@ -1,122 +1,269 @@
+    const str_view details_data_str = make_string_view_strip_quotes(details_data);
+    const str_view details_line_str = make_string_view_strip_quotes(details_line);
+    const str_view waf_unique_id_str = make_string_view_strip_quotes(waf_unique_id);
+
+    return to_json_string({
+            resource_id,
+            operation_name,
+            category,
+            waf_logging::property {
+                instance_id,
+                client_ip,
+                client_port,
+                request_uri,
+                rule_set_type,
+                rule_set_version,
+                rule_id_str,
+                message_str,
+                action_str,
+                site_str,
+                waf_logging::detail {
+                    details_messages,
+                    details_data_str,
+                    details_file_str,
+                    details_line_str
+                },
+            hostname,
+            waf_unique_id_str
+        }
+    });
 }
 
-// Main function:  get fields from modsec, set the protobuf object and write to file in json.
-int generate_json(char** result_json, char* resourceId, char* operationName, char* category, char* instanceId, char* clientIP, char* clientPort, const char* requestUri, char* ruleSetType, char* ruleSetVersion, char* ruleId, char* messages, int action, int site, char* details_messages, char* details_data, char* details_file, char* details_line, const char* hostname, char* waf_unique_id) {
-    string json_string;
-    char* json_str;
+// Get fields from ModSec and format them to JSON.
+char* generate_json(const char* resource_id, const char* operation_name, const char* category,
+        const char* instance_id, const char* client_ip, const char* client_port, const char* request_uri,
+        const char* rule_set_type, const char* rule_set_version, const char* rule_id, const char* messages,
+        const int action, const int site, const char* details_messages, const char* details_data,
+        const char* details_file, const char* details_line, const char* hostname, const char* waf_unique_id) {
     
     try {
-        // Set Waf format.
-        json_string = get_json_log_message(resourceId, operationName, category, instanceId, clientIP, clientPort, requestUri, ruleSetType, ruleSetVersion, ruleId, messages, action, site, details_messages, details_data, details_file, details_line, hostname, waf_unique_id); 
-    }
-    catch (...) {
-        return WAF_LOG_UTIL_FAILED;
-    }
+        const std::string json_string = get_json_log_message(resource_id, operation_name, category,
+                instance_id, client_ip, client_port, request_uri, rule_set_type, rule_set_version,
+                rule_id, messages, action, site, details_messages, details_data, details_file,
+                details_line, hostname, waf_unique_id);
+        char* result = static_cast<char*>(std::malloc(json_string.length() + 2)); // Two extra bytes for \n and \0
```

Comment:
might need to test E2E with td-agent parser since we are using regex parser, or just make sure the log output is the same before/after the change

Reply:
Yes, we'll definitely do E2E once the tenant changes are also implemented and I pull in the ModSecurity version that has this fix.